rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,258 @@
+
+## Detailed Design or Proposal
+
+<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->
+
+### Hook based architecture.
+
+Generally a component will have the following different files. Let us consider a `Sample` component
+
+- Sample.tsx
+- useSample.ts
+- useSampleClasses/useSampleStyles.ts
+- renderSample.tsx
+- Sample.types.ts
+
+#### Sample.tsx
+
+The 'final' product, simply forwards a ref and uses all the previous building blocks to expose the final component
+
+```typescript
+export const Sample = React.forwardRef<HTMLElement, SampleProps>((props, ref) => {
+  const state = useSample(props, ref);
+  useSampleStyles(state);
+
+  return renderSample(state);
+});
+```
+
+#### useSample.ts
+
+Accepts the component props and handles the internal state that the component might need.
+
+`State` here can be pretty broad, you could also consume context or create effects. This hook should be what the component relies on to function/render
+
+```typescript
+import { makeMergeProps, resolveShorthandProps } from '@fluentui/react-utils';
+
+/**
+ * Defines the different slots that can be rendered in this component
+ *
+ * This means that an icon can be rendered 'somewhere'
+ */
+export const sampleShorthandProps = ['icon'];
+
+// Creates a helper function to merge props while respecting slot props
+const mergeProps = makeMergeProps({ deepMerge: sampleShorthandProps });
+
+/**
+ * @parameter props -> these are normal React props for JSX components
+ * @parameter ref -> In case someone wants a ref to the root DOM element
+ * @parameter defaultProps -> safe default props, we set defaults internally but if consumer wants component variants with different defaults, they can set them
+ */
+export const useSample = (props: SampleProps, ref: React.Ref<HTMLElement>, defaultProps?: SampleProps): SampleState => {
+  const resolvedRef = ref || React.useRef();
+
+  // merges the props we declare internally and what is passed in
+  // by a consumer
+  const mergedProps = mergeProps(
+    {
+      // This is essentially shorthand for the `root` slot
+      ref: resolvedRef,
+      as: 'div',
+      icon: { as: 'span' },
+    },
+    defaultProps,
+    resolveShorthandProps(props, sampleShorthandProps),
+  );
+
+  if (checkedValues || onCheckedValuesChange) {
+    mergedProps.hasCheckMark = true;
+  }
+
+  // Anything else the component needs to manage its own state and render
+  [someState, setSomeState] = React.useState();
+  const { contextValue } = React.useContext();
+  React.useEffect(...);
+
+  // make an 'uber' state
+  const state = { someState, contextValue, ...mergedProps}
+
+  return state;
+};
+```
+
+#### useSampleClasses/useSampleStyles.ts
```

Comment:
```suggestion
#### useSampleClasses.ts
```

Reply:
hmmmm `makeStyles` then `useSampleClasses` ? seems to not be consistent for me

Reply:
Let's update to `useComponentStyles`, no `classes` since it causes confusion with `makeStyles`

Reply:
updated to only be `useSampleStyles`